plugins/reindex/src/main/java/org/elasticsearch/plugin/reindex/AbstractAsyncBulkByScrollAction.java

```diff
@@ -0,0 +1,352 @@
+                    total.set(response.getHits().getTotalHits());
+                    onScrollResponse(response);
+                }
+
+                @Override
+                public void onFailure(Throwable e) {
+                    logger.error("failed while executing the initial scroll request", e);
+                    listener.onFailure(e);
+                }
+            });
+        } catch (Throwable t) {
+            finishHim(t);
+        }
+    }
+
+    void onScrollResponse(SearchResponse searchResponse) {
+        threadPool.executor(ThreadPool.Names.GENERIC).execute(() -> {
+            try {
+                scroll.set(searchResponse.getScrollId());
+                SearchHit[] docs = searchResponse.getHits().getHits();
+                logger.debug("scroll returned [{}] documents with a scroll id of [{}]", docs.length, searchResponse.getScrollId());
+                if (docs.length == 0) {
+                    startNormalTermination();
+                    return;
+                }
+                batches.incrementAndGet();
+                List<SearchHit> docsIterable = Arrays.asList(docs);
+                if (mainRequest.size() != -1) {
+                    // Truncate the docs if we have more than the request size
+                    long remaining = max(0, mainRequest.size() - successfullyProcessed());
+                    if (remaining <= docs.length) {
+                        if (remaining < docs.length) {
+                            docsIterable = docsIterable.subList(0, (int) remaining);
+                        }
+                    }
+                }
+                BulkRequest request = buildBulk(docsIterable);
+                if (request.requests().isEmpty()) {
+                    /*
+                     * If we noop-ed the entire batch then just skip to the next
+                     * batch or the BulkRequest would fail validation.
+                     */
+                    startNextScrollRequest();
+                    return;
+                }
+                request.timeout(mainRequest.timeout());
+                request.consistencyLevel(mainRequest.consistency());
+                if (logger.isDebugEnabled()) {
+                    logger.debug("sending [{}] entry, [{}] bulk request", request.requests().size(),
+                            new ByteSizeValue(request.estimatedSizeInBytes()));
+                }
+                client.bulk(request, new ActionListener<BulkResponse>() {
+                    @Override
+                    public void onResponse(BulkResponse response) {
+                        onBulkResponse(response);
+                    }
+
+                    @Override
+                    public void onFailure(Throwable e) {
+                        finishHim(e);
+                    }
+                });
+            } catch (Throwable t) {
+                finishHim(t);
+            }
+        });
+    }
+
+    void onBulkResponse(BulkResponse response) {
```

Comment:
Right. Its the listener usage that makes me use package private method. If you don't java creates an `access$` method so the listener doesn't have to break the privacy rules. Elasticsearch (and Lucene) tend to declare the method package private rather than let the compiler make the synthetic method. Its not clear to me which one is better, but I _think_ its more common in Elasticsearch to do it this way. See `TransportBroadcastAction`, for example.